File: reconcile.js
#!/usr/bin/env node
/**
 * 仓库渠道数据 vs NC系统销售出库 自动对账脚本
 * 匹配逻辑: 物料编码 + 物流单号/快递单号
 */
const fs = require('fs');

function parseAnnotatedCSV(annotatedCsv) {
    const rows = [];
    const lines = annotatedCsv.trim().split('\n');
    let currentRow = null;
    let currentData = '';

    for (const line of lines) {
        const match = line.match(/^\[row=(\d+)\] (.*)/);
        if (match) {
            if (currentRow !== null) {
                rows.push({ row: currentRow, data: currentData });
            }
            currentRow = parseInt(match[1]);
            currentData = match[2];
        } else {
            currentData += '\n' + line;
        }
    }
    if (currentRow !== null) {
        rows.push({ row: currentRow, data: currentData });
    }

    // Parse CSV for each row
    return rows.map(r => {
        const fields = parseCSVLine(r.data);
        return { row: r.row, fields };
    });
}

function parseCSVLine(line) {
    const result = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];
        if (inQuotes) {
            if (ch === '"') {
                if (i + 1 < line.length && line[i + 1] === '"') {
                    current += '"';
                    i++;
                } else {
                    inQuotes = false;
                }
            } else {
                current += ch;
            }
        } else {
            if (ch === '"') {
                inQuotes = true;
            } else if (ch === ',') {
                result.push(current);
                current = '';
            } else {
                current += ch;
            }
        }
    }
    result.push(current);
    return result;
}

function parseNumber(str) {
    if (!str || str.trim() === '') return 0;
    const cleaned = str.replace(/,/g, '').trim();
    const num = parseFloat(cleaned);
    return isNaN(num) ? 0 : num;
}

function main() {
    const warehouseFile = process.env.USERPROFILE + '/.claude/projects/C--Users-Administrator-Desktop-Claude-code/5ac0f536-086a-4e50-9474-cea2f8515324/tool-results/bx9gr2d6o.txt';
    const ncFile = process.env.USERPROFILE + '/.claude/projects/C--Users-Administrator-Desktop-Claude-code/5ac0f536-086a-4e50-9474-cea2f8515324/tool-results/bp32qfusy.txt';

    console.log('Loading warehouse data...');
    const whRaw = JSON.parse(fs.readFileSync(warehouseFile, 'utf-8'));
    const whRows = parseAnnotatedCSV(whRaw.data.annotated_csv);
    console.log(`  Loaded ${whRows.length} rows (including header)`);

    console.log('Loading NC data...');
    const ncRaw = JSON.parse(fs.readFileSync(ncFile, 'utf-8'));
    const ncRows = parseAnnotatedCSV(ncRaw.data.annotated_csv);
    console.log(`  Loaded ${ncRows.length} rows (including header)`);

    // Separate header and data
    const whHeader = whRows[0];
    const whData = whRows.slice(1);
    const ncHeader = ncRows[0];
    const ncData = ncRows.slice(1);

    // Build NC indexes
    // Key: material_code + tracking_no
    const ncIndex = new Map();
    // Key: material_code only
    const ncByMaterial = new Map();

    for (const { row, fields } of ncData) {
        const materialCode = (fields[13] || '').trim();
        const trackingNo = (fields[26] || '').trim();
        const qtyOut1 = (fields[31] || '').trim();
        const qtyOut2 = (fields[32] || '').trim();

        const record = {
            row,
            materialCode,
            materialName: (fields[14] || '').trim(),
            customer: (fields[11] || '').trim(),
            docNo: (fields[5] || '').trim(),
            trackingNo,
            logistics: (fields[25] || '').trim(),
            qtyOut1: parseNumber(qtyOut1),
            qtyOut2: parseNumber(qtyOut2),
            date: (fields[1] || '').trim(),
            rawQty1: qtyOut1,
            rawQty2: qtyOut2,
        };

        const key = `${materialCode}|||${trackingNo}`;
        if (!ncIndex.has(key)) ncIndex.set(key, []);
        ncIndex.get(key).push(record);

        if (!ncByMaterial.has(materialCode)) ncByMaterial.set(materialCode, []);
        ncByMaterial.get(materialCode).push(record);
    }

    console.log(`\nNC index: ${ncIndex.size} unique (material+tracking) keys`);
    console.log(`NC material-only index: ${ncByMaterial.size} unique materials`);

    // Reconciliation
    const results = [];
    let matchedCount = 0;
    let unmatchedCount = 0;
    let partialMatchCount = 0;

    for (const { row, fields } of whData) {
        const channel = (fields[0] || '').trim();
        const date = (fields[1] || '').trim();
        const logistics = (fields[2] || '').trim();
        const trackingNo = (fields[3] || '').trim();
        const customer = (fields[4] || '').trim();
        const entity = (fields[5] || '').trim();
        const project = (fields[6] || '').trim();
        const materialCode = (fields[7] || '').trim();
        const materialName = (fields[8] || '').trim();
        const whQtyRaw = (fields[9] || '').trim();
        const poNo = (fields[10] || '').trim();
        const remark = (fields[11] || '').trim();
        const channel2 = (fields[12] || '').trim();
        const productLine = (fields[13] || '').trim();

        const whQty = parseNumber(whQtyRaw);

        const result = {
            whRow: row,
            channel,
            date,
            logistics,
            trackingNo,
            customer,
            materialCode,
            materialName,
            whQty,
            whQtyRaw,
            poNo,
            remark,
            channel2,
            productLine,
            ncMatches: [],
            ncTotalQty: 0,
            matchStatus: '',
            qtyDiff: 0,
        };

        const key = `${materialCode}|||${trackingNo}`;

        if (ncIndex.has(key)) {
            const ncRecords = ncIndex.get(key);
            let ncTotal = 0;
            for (const nc of ncRecords) {
                ncTotal += nc.qtyOut1;
                result.ncMatches.push(nc);
            }
            result.ncTotalQty = ncTotal;
            result.qtyDiff = whQty - ncTotal;

            if (Math.abs(result.qtyDiff) < 0.01) {
                result.matchStatus = '✅ 匹配一致';
                matchedCount++;
            } else {
                result.matchStatus = '⚠️ 数量差异';
                partialMatchCount++;
            }
        } else if (ncByMaterial.has(materialCode)) {
            const ncRecords = ncByMaterial.get(materialCode);
            let ncTotal = 0;
            for (const nc of ncRecords) {
                ncTotal += nc.qtyOut1;
                result.ncMatches.push(nc);
            }
            result.ncTotalQty = ncTotal;
            result.qtyDiff = whQty - ncTotal;

            if (Math.abs(result.qtyDiff) < 0.01) {
                result.matchStatus = '✅ 物料匹配一致(单号不同)';
                matchedCount++;
            } else {
                result.matchStatus = '🔶 物料匹配-数量差异(单号不同)';
                partialMatchCount++;
            }
        } else {
            result.matchStatus = '❌ NC无匹配';
            unmatchedCount++;
        }

        results.push(result);
    }

    // Find NC records not matched
    const matchedNcRows = new Set();
    for (const r of results) {
        for (const nc of r.ncMatches) {
            matchedNcRows.add(nc.row);
        }
    }

    const ncUnmatched = [];
    for (const { row, fields } of ncData) {
        if (!matchedNcRows.has(row)) {
            ncUnmatched.push({
                row,
                materialCode: (fields[13] || '').trim(),
                materialName: (fields[14] || '').trim(),
                customer: (fields[11] || '').trim(),
                docNo: (fields[5] || '').trim(),
                trackingNo: (fields[26] || '').trim(),
                qtyOut1: (fields[31] || '').trim(),
                logistics: (fields[25] || '').trim(),
                date: (fields[1] || '').trim(),
            });
        }
    }

    // Print summary
    console.log(`\n========== 对账结果汇总 ==========`);
    console.log(`仓库发货记录总数: ${results.length}`);
    console.log(`  ✅ 匹配一致: ${matchedCount}`);
    console.log(`  ⚠️ 数量差异: ${partialMatchCount}`);
    console.log(`  ❌ NC无匹配: ${unmatchedCount}`);
    console.log(`NC未匹配记录数: ${ncUnmatched.length}`);

    // Quantity differences
    const diffResults = results.filter(r => r.matchStatus.includes('差异'));
    console.log(`\n========== 数量差异明细 (${diffResults.length}条) ==========`);
    for (const r of diffResults.slice(0, 30)) {
        console.log(`  [${r.materialCode}] ${r.materialName.substring(0, 50)}`);
        console.log(`    仓库: ${r.whQty} | NC: ${r.ncTotalQty} | 差异: ${r.qtyDiff} | 单号: ${r.trackingNo}`);
    }

    // Unmatched warehouse
    const unmatchedWH = results.filter(r => r.matchStatus === '❌ NC无匹配');
    console.log(`\n========== 仓库有但NC无匹配 (${unmatchedWH.length}条) ==========`);
    for (const r of unmatchedWH.slice(0, 20)) {
        console.log(`  [${r.materialCode}] ${r.materialName.substring(0, 50)} | Qty:${r.whQty} | 单号:${r.trackingNo} | ${r.customer}`);
    }

    // Unmatched NC
    console.log(`\n========== NC有但仓库无匹配 (${ncUnmatched.length}条) ==========`);
    for (const r of ncUnmatched.slice(0, 20)) {
        console.log(`  [${r.materialCode}] ${r.materialName.substring(0, 50)} | Qty:${r.qtyOut1} | 单号:${r.trackingNo} | ${r.customer}`);
    }

    // Save JSON output
    const output = {
        summary: {
            warehouseTotal: results.length,
            matched: matchedCount,
            quantityDiff: partialMatchCount,
            unmatched: unmatchedCount,
            ncUnmatched: ncUnmatched.length,
        },
        diffDetails: diffResults.map(r => ({
            materialCode: r.materialCode,
            materialName: r.materialName,
            trackingNo: r.trackingNo,
            customer: r.customer,
            whQty: r.whQty,
            ncQty: r.ncTotalQty,
            diff: r.qtyDiff,
            status: r.matchStatus,
            whRow: r.whRow,
            ncTrackingNos: [...new Set(r.ncMatches.map(n => n.trackingNo))],
            ncDocNos: [...new Set(r.ncMatches.map(n => n.docNo))],
        })),
        unmatchedWarehouse: unmatchedWH.map(r => ({
            materialCode: r.materialCode,
            materialName: r.materialName,
            trackingNo: r.trackingNo,
            logistics: r.logistics,
            customer: r.customer,
            whQty: r.whQty,
            whRow: r.whRow,
        })),
        unmatchedNC: ncUnmatched.map(r => ({
            materialCode: r.materialCode,
            materialName: r.materialName,
            trackingNo: r.trackingNo,
            customer: r.customer,
            ncQty: r.qtyOut1,
            ncRow: r.row,
            docNo: r.docNo,
        })),
    };

    const outputPath = 'C:/Users/Administrator/Desktop/Claude code/reconcile_result.json';
    fs.writeFileSync(outputPath, JSON.stringify(output, null, 2), 'utf-8');
    console.log(`\n详细结果已保存至: ${outputPath}`);

    // Generate CSV for writing to Feishu
    generateReconciliationCSV(results, ncUnmatched);
}

function generateReconciliationCSV(results, ncUnmatched) {
    // Generate the reconciliation sheet content as CSV rows
    const csvRows = [];

    // Section 1: Summary
    csvRows.push(['对账结果汇总']);
    csvRows.push(['项目', '数值']);
    const totalWH = results.length;
    const matched = results.filter(r => r.matchStatus.includes('一致') && !r.matchStatus.includes('差异')).length;
    const qtyDiff = results.filter(r => r.matchStatus.includes('差异')).length;
    const unmatched = results.filter(r => r.matchStatus === '❌ NC无匹配').length;
    csvRows.push(['仓库发货总记录', totalWH]);
    csvRows.push(['匹配一致', matched]);
    csvRows.push(['数量差异', qtyDiff]);
    csvRows.push(['NC无匹配', unmatched]);
    csvRows.push(['NC未匹配记录', ncUnmatched.length]);
    csvRows.push([]);

    // Section 2: Quantity Differences
    csvRows.push(['数量差异明细']);
    csvRows.push(['状态', '渠道', '出货日期', '物料编码', '物料名称', '物流单号', '客户', '仓库数量', 'NC出库数量', '差异', 'NC单号', '仓库行号']);
    const diffResults = results.filter(r => r.matchStatus.includes('差异'));
    for (const r of diffResults) {
        csvRows.push([
            r.matchStatus,
            r.channel,
            r.date,
            r.materialCode,
            r.materialName,
            r.trackingNo,
            r.customer,
            r.whQty,
            r.ncTotalQty,
            r.qtyDiff,
            [...new Set(r.ncMatches.map(n => n.docNo))].join('; '),
            r.whRow,
        ]);
    }
    csvRows.push([]);

    // Section 3: Warehouse unmatched
    csvRows.push(['仓库有但NC无匹配']);
    csvRows.push(['渠道', '出货日期', '物料编码', '物料名称', '物流单号', '物流公司', '客户', '数量', '仓库行号']);
    const unmatchedWH = results.filter(r => r.matchStatus === '❌ NC无匹配');
    for (const r of unmatchedWH) {
        csvRows.push([
            r.channel,
            r.date,
            r.materialCode,
            r.materialName,
            r.trackingNo,
            r.logistics,
            r.customer,
            r.whQty,
            r.whRow,
        ]);
    }
    csvRows.push([]);

    // Section 4: NC unmatched
    csvRows.push(['NC有但仓库无匹配']);
    csvRows.push(['物料编码', '物料名称', '快递单号', '物流公司', '客户', 'NC出库数量', '单据号', 'NC行号']);
    for (const r of ncUnmatched) {
        csvRows.push([
            r.materialCode,
            r.materialName,
            r.trackingNo,
            r.logistics,
            r.customer,
            r.qtyOut1,
            r.docNo,
            r.row,
        ]);
    }

    // Write CSV
    const csvContent = csvRows.map(row =>
        row.map(cell => {
            const str = String(cell);
            // Quote if contains comma, newline, or quote
            if (str.includes(',') || str.includes('\n') || str.includes('"')) {
                return '"' + str.replace(/"/g, '""') + '"';
            }
            return str;
        }).join(',')
    ).join('\n');

    const csvPath = 'C:/Users/Administrator/Desktop/Claude code/reconcile_output.csv';
    fs.writeFileSync(csvPath, csvContent, 'utf-8');
    console.log(`CSV输出已保存至: ${csvPath}`);
    console.log(`CSV行数: ${csvRows.length}`);
}

main();

File: reconcile.py
#!/usr/bin/env python3
"""
仓库渠道数据 vs NC系统销售出库 自动对账脚本
匹配逻辑: 物料编码 + 物流单号/快递单号
"""
import json
import csv
import io
import re
from collections import defaultdict

def parse_annotated_csv(annotated_csv):
    """Parse the annotated CSV format from Feishu sheets"""
    rows = []
    # Remove [row=N] prefix from each logical line
    # Handle CSV properly with quoted fields
    lines = annotated_csv.strip().split('\n')
    current_row = None
    current_data = ""

    for line in lines:
        match = re.match(r'^\[row=(\d+)\] (.*)', line)
        if match:
            # Save previous row if exists
            if current_row is not None:
                rows.append((current_row, current_data))
            current_row = int(match.group(1))
            current_data = match.group(2)
        else:
            # Continuation of previous row (multi-line field)
            current_data += '\n' + line

    if current_row is not None:
        rows.append((current_row, current_data))

    # Parse CSV rows
    parsed = []
    for row_num, data in rows:
        reader = csv.reader(io.StringIO(data))
        try:
            fields = next(reader)
            parsed.append((row_num, fields))
        except StopIteration:
            pass

    return parsed

def load_data(filepath):
    """Load data from Feishu API JSON output file"""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    data = json.loads(content)
    annotated_csv = data['data']['annotated_csv']
    col_indices = data['data']['col_indices']

    rows = parse_annotated_csv(annotated_csv)
    return rows, col_indices

def main():
    # File paths
    warehouse_file = r"C:\Users\Administrator\.claude\projects\C--Users-Administrator-Desktop-Claude-code\5ac0f536-086a-4e50-9474-cea2f8515324\tool-results\bx9gr2d6o.txt"
    nc_file = r"C:\Users\Administrator\.claude\projects\C--Users-Administrator-Desktop-Claude-code\5ac0f536-086a-4e50-9474-cea2f8515324\tool-results\bp32qfusy.txt"

    print("Loading warehouse data...")
    warehouse_rows, wh_cols = load_data(warehouse_file)
    print(f"  Loaded {len(warehouse_rows)} rows (including header)")

    print("Loading NC data...")
    nc_rows, nc_cols = load_data(nc_file)
    print(f"  Loaded {len(nc_rows)} rows (including header)")

    # Parse warehouse data
    # Columns: A=渠道, B=出货日期, C=物流公司, D=物流单号, E=收件公司, F=出货主体
    #           G=项目名称, H=物料编码, I=物料名称, J=数量, K=采购订单号
    #           L=备注, M=渠道, N=产品线, O=备注
    wh_header = warehouse_rows[0][1]
    wh_data = warehouse_rows[1:]  # Skip header

    # Parse NC data
    # Columns: N=物料编码(13), O=物料名称(14), L=客户名称(11)
    #           Z=物流公司(25), AA=快递单号(26), F=单据号(5)
    #           AF=出库(31), AG=出库(32), B=业务日期(1)
    nc_header = nc_rows[0][1]
    nc_data = nc_rows[1:]  # Skip header

    # Build NC index: key = (物料编码, 快递单号) -> list of records
    nc_index = defaultdict(list)
    for row_num, fields in nc_data:
        if len(fields) > 26:
            material_code = fields[13].strip() if len(fields) > 13 else ""  # N column
            tracking_no = fields[26].strip() if len(fields) > 26 else ""    # AA column
            qty_out1 = fields[31].strip() if len(fields) > 31 else ""       # AF column
            qty_out2 = fields[32].strip() if len(fields) > 32 else ""       # AG column

            key = (material_code, tracking_no)
            nc_index[key].append({
                'row': row_num,
                'material_code': material_code,
                'material_name': fields[14].strip() if len(fields) > 14 else "",
                'customer': fields[11].strip() if len(fields) > 11 else "",
                'doc_no': fields[5].strip() if len(fields) > 5 else "",
                'tracking_no': tracking_no,
                'logistics': fields[25].strip() if len(fields) > 25 else "",
                'qty_out1': qty_out1,
                'qty_out2': qty_out2,
                'date': fields[1].strip() if len(fields) > 1 else "",
            })

    # Also build NC index by material code only
    nc_by_material = defaultdict(list)
    for row_num, fields in nc_data:
        if len(fields) > 14:
            material_code = fields[13].strip()
            tracking_no = fields[26].strip() if len(fields) > 26 else ""
            qty_out1 = fields[31].strip() if len(fields) > 31 else ""
            qty_out2 = fields[32].strip() if len(fields) > 32 else ""

            nc_by_material[material_code].append({
                'row': row_num,
                'material_code': material_code,
                'material_name': fields[14].strip() if len(fields) > 14 else "",
                'customer': fields[11].strip() if len(fields) > 11 else "",
                'doc_no': fields[5].strip() if len(fields) > 5 else "",
                'tracking_no': tracking_no,
                'logistics': fields[25].strip() if len(fields) > 25 else "",
                'qty_out1': qty_out1,
                'qty_out2': qty_out2,
                'date': fields[1].strip() if len(fields) > 1 else "",
            })

    print(f"\nNC index built: {len(nc_index)} unique (material+tracking) keys")
    print(f"NC material-only index: {len(nc_by_material)} unique materials")

    # Reconciliation results
    results = []
    matched_count = 0
    unmatched_count = 0
    partial_match_count = 0

    for row_num, fields in wh_data:
        if len(fields) < 10:
            continue

        channel = fields[0].strip() if len(fields) > 0 else ""
        date = fields[1].strip() if len(fields) > 1 else ""
        logistics = fields[2].strip() if len(fields) > 2 else ""
        tracking_no = fields[3].strip() if len(fields) > 3 else ""
        customer = fields[4].strip() if len(fields) > 4 else ""
        entity = fields[5].strip() if len(fields) > 5 else ""
        project = fields[6].strip() if len(fields) > 6 else ""
        material_code = fields[7].strip() if len(fields) > 7 else ""
        material_name = fields[8].strip() if len(fields) > 8 else ""
        wh_qty = fields[9].strip() if len(fields) > 9 else ""
        po_no = fields[10].strip() if len(fields) > 10 else ""
        remark = fields[11].strip() if len(fields) > 11 else ""

        # Try to parse warehouse quantity
        try:
            wh_qty_num = float(wh_qty.replace(',', '')) if wh_qty else 0
        except ValueError:
            wh_qty_num = 0

        # Match by material_code + tracking_no
        key = (material_code, tracking_no)

        result = {
            'wh_row': row_num,
            'channel': channel,
            'date': date,
            'logistics': logistics,
            'tracking_no': tracking_no,
            'customer': customer,
            'material_code': material_code,
            'material_name': material_name,
            'wh_qty': wh_qty_num,
            'po_no': po_no,
            'remark': remark,
            'nc_matches': [],
            'nc_total_qty': 0,
            'match_status': '',
            'qty_diff': 0,
        }

        if key in nc_index:
            nc_records = nc_index[key]
            nc_total = 0
            for nc in nc_records:
                try:
                    nc_qty = float(nc['qty_out1'].replace(',', '')) if nc['qty_out1'] else 0
                except ValueError:
                    nc_qty = 0
                nc_total += nc_qty
                result['nc_matches'].append(nc)

            result['nc_total_qty'] = nc_total
            result['qty_diff'] = wh_qty_num - nc_total

            if abs(result['qty_diff']) < 0.01:
                result['match_status'] = '✅ 匹配一致'
                matched_count += 1
            else:
                result['match_status'] = '⚠️ 数量差异'
                partial_match_count += 1
        else:
            # Try matching by material code only (fallback)
            if material_code in nc_by_material:
                nc_records = nc_by_material[material_code]
                nc_total = 0
                for nc in nc_records:
                    try:
                        nc_qty = float(nc['qty_out1'].replace(',', '')) if nc['qty_out1'] else 0
                    except ValueError:
                        nc_qty = 0
                    nc_total += nc_qty
                    result['nc_matches'].append(nc)

                result['nc_total_qty'] = nc_total
                result['qty_diff'] = wh_qty_num - nc_total
                result['match_status'] = '🔶 仅物料匹配(单号不同)'
                partial_match_count += 1
            else:
                result['match_status'] = '❌ NC无匹配'
                unmatched_count += 1

        results.append(result)

    # Also find NC records not matched to any warehouse record
    matched_nc_rows = set()
    for r in results:
        for nc in r['nc_matches']:
            matched_nc_rows.add(nc['row'])

    nc_unmatched = []
    for row_num, fields in nc_data:
        if row_num not in matched_nc_rows and len(fields) > 14:
            nc_unmatched.append({
                'row': row_num,
                'material_code': fields[13].strip() if len(fields) > 13 else "",
                'material_name': fields[14].strip() if len(fields) > 14 else "",
                'customer': fields[11].strip() if len(fields) > 11 else "",
                'doc_no': fields[5].strip() if len(fields) > 5 else "",
                'tracking_no': fields[26].strip() if len(fields) > 26 else "",
                'qty_out1': fields[31].strip() if len(fields) > 31 else "",
            })

    # Print summary
    print(f"\n========== 对账结果汇总 ==========")
    print(f"仓库发货记录总数: {len(results)}")
    print(f"  ✅ 匹配一致: {matched_count}")
    print(f"  ⚠️ 数量差异: {partial_match_count}")
    print(f"  ❌ NC无匹配: {unmatched_count}")
    print(f"NC未匹配记录数: {len(nc_unmatched)}")

    # Print quantity differences
    print(f"\n========== 数量差异明细 ==========")
    diff_results = [r for r in results if r['match_status'] == '⚠️ 数量差异']
    for r in diff_results[:20]:
        print(f"  [{r['material_code']}] {r['material_name'][:40]}")
        print(f"    仓库: {r['wh_qty']} | NC: {r['nc_total_qty']} | 差异: {r['qty_diff']}")
        print(f"    物流单号: {r['tracking_no']} | 客户: {r['customer']}")

    if len(diff_results) > 20:
        print(f"  ... 还有 {len(diff_results) - 20} 条差异")

    # Print unmatched
    print(f"\n========== 仓库有但NC无匹配 (前20条) ==========")
    unmatched_results = [r for r in results if r['match_status'] == '❌ NC无匹配']
    for r in unmatched_results[:20]:
        print(f"  [{r['material_code']}] {r['material_name'][:50]} | Qty:{r['wh_qty']} | 单号:{r['tracking_no']}")

    if len(unmatched_results) > 20:
        print(f"  ... 还有 {len(unmatched_results) - 20} 条未匹配")

    print(f"\n========== NC有但仓库无匹配 (前20条) ==========")
    for r in nc_unmatched[:20]:
        print(f"  [{r['material_code']}] {r['material_name'][:50]} | Qty:{r['qty_out1']} | 单号:{r['tracking_no']}")

    if len(nc_unmatched) > 20:
        print(f"  ... 还有 {len(nc_unmatched) - 20} 条未匹配")

    # Save results to JSON for later use
    output = {
        'summary': {
            'warehouse_total': len(results),
            'matched': matched_count,
            'quantity_diff': partial_match_count,
            'unmatched': unmatched_count,
            'nc_unmatched': len(nc_unmatched),
        },
        'diff_details': [
            {
                'material_code': r['material_code'],
                'material_name': r['material_name'],
                'tracking_no': r['tracking_no'],
                'customer': r['customer'],
                'wh_qty': r['wh_qty'],
                'nc_qty': r['nc_total_qty'],
                'diff': r['qty_diff'],
                'status': r['match_status'],
                'wh_row': r['wh_row'],
            }
            for r in diff_results
        ],
        'unmatched_warehouse': [
            {
                'material_code': r['material_code'],
                'material_name': r['material_name'],
                'tracking_no': r['tracking_no'],
                'customer': r['customer'],
                'wh_qty': r['wh_qty'],
                'wh_row': r['wh_row'],
            }
            for r in unmatched_results
        ],
        'unmatched_nc': [
            {
                'material_code': r['material_code'],
                'material_name': r['material_name'],
                'tracking_no': r['tracking_no'],
                'customer': r['customer'],
                'nc_qty': r['qty_out1'],
                'nc_row': r['row'],
            }
            for r in nc_unmatched
        ],
    }

    output_path = r"C:\Users\Administrator\Desktop\Claude code\reconcile_result.json"
    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print(f"\n详细结果已保存至: {output_path}")

    return results, nc_unmatched, output

if __name__ == '__main__':
    results, nc_unmatched, output = main()
